cpu/atmega2560/periph/gpio.c

```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright (C) 2015 haw-hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     driver_periph
+ * @{
+ *
+ * @file
+ * @brief       Low-level GPIO driver implementation for ATmega2560
+ *
+ * @author      René Herthel <rene-herthel@outlook.de>
+ *
+ * @}
+ */
+
+
+#include <stdio.h>
+
+#include <avr/interrupt.h>
+
+#include "cpu.h"
+#include "periph/gpio.h"
+#include "periph_conf.h"
+
+#define GPIO_BASE_PORT_A        (0x20)
+#define GPIO_OFFSET_PORT_H      (0xCB)
+#define GPIO_OFFSET_PIN_PORT    (0x02)
+#define GPIO_OFFSET_PIN_PIN     (0x03)
+#define GPIO_EXT_INT_NUMOF      (5U)
+
+typedef struct {
+    gpio_cb_t cb;
+    void *arg;
+} gpio_state_t;
+
+static gpio_state_t config[GPIO_EXT_INT_NUMOF];
+
+/**
+ * @brief     Extract the pin number of the given pin
+ */
+static inline uint8_t _pin_num(gpio_t pin)
+{
+    return (pin & 0x0f);
+}
+
+/**
+ * @brief     Extract the port number of the given pin
+ */
+static inline uint8_t _port_num(gpio_t pin)
+{
+    return (pin >> 4) & 0x0f;
+}
+
+/**
+ * @brief     Generate the port address of the give pin.
+ */
+static inline uint16_t _port_addr(gpio_t pin)
+{
+    uint8_t port_num = _port_num(pin);
+    uint16_t port_addr = port_num * GPIO_OFFSET_PIN_PIN;
+
+    port_addr += GPIO_BASE_PORT_A;
+    port_addr += GPIO_OFFSET_PIN_PORT;
+
+    if (port_num > PORT_G) {
+        port_addr += GPIO_OFFSET_PORT_H;
+    }
+
+    return port_addr;
+}
+
+/**
+ * @brief     Extract the data direction register address of the given pin
+ */
+static inline uint8_t _ddr_addr(gpio_t pin)
+{
+    return (_port_addr(pin) - 0x01);
+}
+
+int gpio_init(gpio_t pin, gpio_dir_t dir, gpio_pp_t pullup)
+{
+    int ret_code;
+
+    if (dir == GPIO_DIR_OUT) {
+        _SFR_MEM8(_ddr_addr(pin)) |= (1 << _pin_num(pin));
+        ret_code = bit_is_set(_SFR_MEM8(_ddr_addr(pin)), _pin_num(pin));
+    }
+    else {
+        _SFR_MEM8(_ddr_addr(pin)) &= ~(1 << _pin_num(pin));
+        ret_code = bit_is_clear(_SFR_MEM8(_ddr_addr(pin)), _pin_num(pin));
+    }
+
+    if (ret_code == 0) {
+        return -1;
+    }
+
+    return 0;
+}
+
+int gpio_init_int(gpio_t pin, gpio_pp_t pullup, gpio_flank_t flank,
```

Comment:
@haukepetersen, with the new gpio design we said, that it is up to the user to know his board and **not**  try to initialize "invalid" port/pin combinations. On the atmega2560 there are IIRC just some external interrupt lines. Do you think it makes sense to add a check here if the port/pin **is** valid?